cpp/velox/shuffle/VeloxShuffleWriter.cc

```diff
@@ -167,8 +167,78 @@ std::shared_ptr<arrow::Array> makeBinaryArray(
   return arrow::MakeArray(arrow::ArrayData::Make(type, 1, {nullptr, offsetBuffer, valueBuffer}));
 }
 
+inline void writeInt64(std::shared_ptr<arrow::Buffer> buffer, int64_t& offset, int64_t value) {
+  memcpy(buffer->mutable_data() + offset, &value, sizeof(int64_t));
+  offset += sizeof(int64_t);
+}
+
+int64_t getMaxCompressedBufferSize(
+    const std::vector<std::shared_ptr<arrow::Buffer>>& buffers,
+    arrow::util::Codec* codec) {
+  int64_t totalSize = 0;
+  for (auto& buffer : buffers) {
+    if (buffer != nullptr) {
+      totalSize += codec->MaxCompressedLen(buffer->size(), nullptr);
+    }
+  }
+  return totalSize;
+}
+
+std::shared_ptr<arrow::RecordBatch> makeCompressedRecordBatch(
+    uint32_t numRows,
+    const std::vector<std::shared_ptr<arrow::Buffer>>& buffers,
+    const std::shared_ptr<arrow::Schema> compressWriteSchema,
+    ShuffleBufferPool* pool,
+    arrow::util::Codec* codec) {
+  std::vector<std::shared_ptr<arrow::Array>> arrays;
+  // header col, numRows
+  {
+    std::shared_ptr<arrow::Buffer> headerBuffer;
+    GLUTEN_THROW_NOT_OK(pool->allocate(headerBuffer, sizeof(uint32_t)));
+    memcpy(headerBuffer->mutable_data(), &numRows, sizeof(uint32_t));
+    int32_t compressType = static_cast<int32_t>(codec->compression_type());
+    memcpy(headerBuffer->mutable_data() + sizeof(uint32_t), &compressType, sizeof(int32_t));
+    arrays.emplace_back(makeBinaryArray(compressWriteSchema->field(0)->type(), headerBuffer, pool));
+  }
+
+  // Length buffer layout |buffers.size()|buffer1 unCompressedLength|buffer1 compressedLength| buffer2...
+  std::shared_ptr<arrow::Buffer> lengthBuffer;
+  GLUTEN_THROW_NOT_OK(pool->allocate(lengthBuffer, (buffers.size() * 2 + 1) * sizeof(int64_t)));
+  int64_t offset = 0;
+  writeInt64(lengthBuffer, offset, buffers.size());
+
+  int64_t compressedBufferMaxSize = getMaxCompressedBufferSize(buffers, codec);
+  std::shared_ptr<arrow::ResizableBuffer> valueBuffer;
+  GLUTEN_THROW_NOT_OK(pool->allocateDirectly(valueBuffer, compressedBufferMaxSize));
+  int64_t compressValueOffset = 0;
+  for (auto& buffer : buffers) {
```

Comment:
Have you considered copy buffers into a large continuous buffer, and then codec->Compress the large buffer? The compression time is much shorter, and compressed buffer size may be smaller compared with compressing several small buffers.

Reply:
Yes, I will try to implement copy small buffers to a big buffer and compress them, the small buffer threshold maybe different for each codec, I will test to get an optimal value.
This PR can improve TPCH performance 3%, so it will be merged first.

Reply:
zstd can use multiple buffers as input, so we can avoid the memcpy.

Reply:
for LZ4, let's have a try to copy into large buffer then compress. I doubt it can have better performance.